api/v1/controllers/collectors.js

```diff
@@ -159,44 +214,91 @@ function deregisterCollector(req, res, next) {
  * @param {Function} next - The next middleware function in the stack
  */
 function heartbeat(req, res, next) {
-  // TODO reject if caller's token is not a collector token
+  const authToken = req.headers.authorization;
+  const timestamp = req.body.timestamp;
+  let keyName;
+  let tokenName;
   const retval = {
     collectorConfig: config.collector,
     generatorsAdded: [],
     generatorsDeleted: [],
     generatorsUpdated: [],
   };
 
-  /*
-   * TODO update the lastHeartbeat column for this collector
-   */
-
-  /*
-   * TODO Populate collectorConfig
-   * - look up any changes made to this collector since the last heartbeat
-   */
-
-  /*
-   * TODO populate generatorsAdded
-   * - look up any new generators assigned to this collector since the last
-   *   heartbeat
-   * - reEncrypt context values using reEncryptSGContextValues function
-   */
-
-  /*
-   * TODO populate generatorsDeleted
-   * - look up any generators UNassigned from this collector since the last
-   *   heartbeat
-   */
-
-  /*
-   * TODO populate generatorsUpdated
-   * - for generators which were already assigned to this collector, look up
-   *   any changes made to the generator since the last heartbeat
-   * - reEncrypt context values using reEncryptSGContextValues function
-   */
-
-  res.status(httpStatus.OK).json(retval);
+  //verify token
+  jwtUtil.verifyCollectorToken(req)
+  .then(() => jwtUtil.getTokenDetailsFromRequest(req))
+  .then((details) => tokenName = details.username)
+
+  //find collector object
+  .then(() => u.findByKey(helper, req.swagger.params))
+  .then((o) => {
+    keyName = o.name;
+
+    //validate collector
+    if (keyName !== tokenName) {
+      throw new apiErrors.ForbiddenError({
+        explanation: 'Token does not match the specified collector',
+      });
```

Comment:
Line 234 to 242 is not required. "verifyCollectorToken" does it already and if we end up using using the "request header" feature, this check would already have been done before the headers are added to the request.

Reply:
verifyCollectorToken only checks that the token belongs to any valid collector that exists. This checks that the collector token matches the collector id or name in the path.

Reply:
Maybe I should just update verifyCollectorToken to do this?
Are there any endpoints a collector needs to access that aren't tied to the key in the path?

Reply:
There is a story to make sure that collector token cannot be spoofed. So, there will not be a case where the those 2 are different.